### PR TITLE
CI: make sanity check work and simplify it

### PR DESCRIPTION
**Pull request by Shwetha-Acharya**
This change indicates any sanity issues before starting the build and hence saves us the waiting time of 2m31s

Fixes: #573

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -170,6 +170,12 @@ jobs:
         with:
           go-version: ${{ env.GO_VERSION }}
 
+      - name: Check auto generated files
+        run: |
+          echo "Failing if any auto generated files are updated, checking 'git status'"
+          git --no-pager diff
+          git status --porcelain 2>&1 | tee /dev/stderr | (! read)
```
[the remark this answers is not in the document]
Comment (raghavendra-talur) [reply .github/workflows/ci.yaml:177]: > Also implementing the check in a github workflow is not a good idea, it should
> really be implemented in a separate scripts invoked from the Makefile, so the
> check can run both locally and in the CI.

+1 to this

[the remark this answers is not in the document]
Comment (ShyamsundarR) [reply .github/workflows/ci.yaml:177]: Auto generated files are post a build, hence this check should run after the build and not before.

[the remark this answers is not in the document]
Comment (nirs) [reply .github/workflows/ci.yaml:177]: I assume that go build does not generate files, and they are generated by `make generate`.

So we can add a check doing:

    $ time make generate
bash: line 1: testbin/golangci-lint: No such file or directory
bash: line 1: /home/nsoffer/src/ramen/bin/controller-gen: No such file or directory
go: creating new go.mod: module tmp
Downloading sigs.k8s.io/controller-tools/cmd/controller-gen@v0.9.2
/home/nsoffer/src/ramen/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."

real	0m2.082s
user	0m5.678s
sys	0m0.868s

And then the current code seems to work:

```
$ git diff; echo $?
diff --git a/api/v1alpha1/zz_generated.deepcopy.go b/api/v1alpha1/zz_generated.deepcopy.go
index 7e358b0..828ce39 100644
--- a/api/v1alpha1/zz_generated.deepcopy.go
+++ b/api/v1alpha1/zz_generated.deepcopy.go
@@ -12,13 +12,10 @@ import (
        corev1 "k8s.io/api/core/v1"
        "k8s.io/apimachinery/pkg/apis/meta/v1"
        "k8s.io/apimachinery/pkg/runtime"
 )
 
-func Test() {
-}
-
 // DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
 func (in *ClusterMaintenanceMode) DeepCopyInto(out *ClusterMaintenanceMode) {
        *out = *in
        if in.Conditions != nil {
                in, out := &in.Conditions, &out.Conditions
0
```

```
$ git status --porcelain 2>&1 | tee /dev/stderr | (! read); echo $?
 M api/v1alpha1/zz_generated.deepcopy.go
1
```

The git status command needs a comment explaining what it does and how it works, and
hopefully a more clear implementation that makes the intent of the code clear.

[the remark this answers is not in the document]
Comment (nirs) [reply .github/workflows/ci.yaml:177]: The current job is broken (not sure why)

I added a bad change:
https://github.com/nirs/ramen/commit/a2bd2cb5c8a46f403a1c899f89813cbc66bd2d7f

But the build passes:
https://github.com/nirs/ramen/actions/runs/5389832325/jobs/9784341426

[the remark this answers is not in the document]
Comment (Shwetha-Acharya) [reply .github/workflows/ci.yaml:177]: > The git status command needs a comment explaining what it does and how it works, and hopefully a more clear >implementation that makes the intent of the code clear.
 I was wondering why was there a need to to tee the output of `git status --porcelain 2>&1` to   /dev/stderr? Is it used anywhere?
And then we are using (!read) to eliminate printing of same things twice (by git status and tee) on stdout 
@nirs @ShyamsundarR

[the remark this answers is not in the document]
Comment (nirs) [reply .github/workflows/ci.yaml:177]: I think this code

    git status --porcelain 2>&1 | tee /dev/stderr | (! read)

is trying to do this:

1. run `git status --porcelain` to get list of modified files using machine readable format
1. pipe to `| tee /dev/stderr` so output is written to stderr
1. check if output is empty `(! read)`

I'm not sure why this does not work in the CI (maybe the pipes?), but we can do this instead:

    git diff-index --quiet HEAD

Which return non-zero exit code if there are modified files - the diff above already show the modifications so we don't need to list the files again.

Please try with a bad change like https://github.com/nirs/ramen/commit/a2bd2cb5c8a46f403a1c899f89813cbc66bd2d7f
to verify that it works.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -188,9 +188,9 @@ jobs:
 
       - name: Check auto generated files
         run: |
-          echo "Failing if any auto generated files are updated, checking 'git status'"
+          echo "Failing if any auto generated files are updated"
           git --no-pager diff
-          git status --porcelain 2>&1 | tee /dev/stderr | (! read)
+          git diff-index --quiet HEAD; echo $?
```
Comment (nirs) [inline .github/workflows/ci.yaml:193]: The `; echo $?' is not needed - it is just for debugging locally to see that we get non zero exit code. But it is good that we have it since the code does not work yet:

```
Failing if any auto generated files are updated
0
```

The reason is that `make docker-build` does not generate the generated files. We need to move this job to the linters job.

The step should run:

```
echo "Checking that generated files were not modified"
make generate
git --no-pager diff
git diff-index --quiet HEAD
```

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -44,7 +44,12 @@ jobs:
           sudo pip3 install yamllint
 
       - name: Run linters
-        run: ./hack/pre-commit.sh
+        run: |
+           ./hack/pre-commit.sh
+           echo "Checking that generated files were not modified"
+           make generate
+           git --no-pager diff
+           git diff-index --quiet HEAD
```
Comment (nirs) [inline .github/workflows/ci.yaml:52]: This should be a separate step:

```
- name: "Checking that generated files were not modified"
  run: |
    make generate
    git --no-pager diff
    git diff-index --quiet HEAD
```

This makes it easy to move the step to another job later, keep the code cleaner, and make it easier to understand failures.

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -44,7 +44,12 @@ jobs:
           sudo pip3 install yamllint
 
       - name: Run linters
-        run: ./hack/pre-commit.sh
+        run: |
+           ./hack/pre-commit.sh
+           echo "Checking that generated files were not modified"
+           make generate
```
Comment (ShyamsundarR) [inline .github/workflows/ci.yaml:50]: I think we need `go mod tidy` as well here, to detect changes to go.mod and go.sum files that were not part of the PR.

In addition `make manifests` maybe required to ensure generated YAML files are part of the PR

**Comment by Shwetha-Acharya** [comment]
Removed the dummy function used for testing. Last build with the test function: https://github.com/Shwetha-Acharya/ramen/actions/runs/5551310804

**Review by nirs — APPROVED** [review]
Looks good, but the commit message miss an important fact - this change fix the check that never worked!

**Comment by Shwetha-Acharya** [comment]
> Looks good, but the commit message miss an important fact - this change fix the check that never worked!

Updated the commit message!

**Comment by ShyamsundarR** [comment]
> Looks good, but the commit message miss an important fact - this change fix the check that never worked!

It would be accurate to state it stopped working post the commit 6c1fc0dad337acb45dc0ac73737ad9586fdb8405 in case we are making this statement in the commit message.

**Review by nirs — APPROVED** [review]
Linking to commit that broke the issue would be nice, see @ShyamsundarR  comment. Not critical for this kind of fix.

**Review by ShyamsundarR — APPROVED** [review]
